aten/src/ATen/record_function.cpp

```diff
@@ -222,6 +233,38 @@ double RecordFunctionCallback::sample_zero_one() {
   return dist(*gen);
 }
 
+bool RecordFunctionCallback::shouldRun(RecordScope scope) const {
+  // first check whether this callback is interested in
+  // the given scope type
+  if (!checkScope(scope)) {
+    return false;
+  }
+  // if we have registered should_run_ function, use it
+  if (should_run_) {
+    return should_run_(*this);
+  }
+  // otherwise potentially do the uniform sampling
+  if (sampling_prob_ != 1.0) {
+    // model the low probability events as events happening
+    // with prob. kLowProb followed by another sampling with
+    // prob. (sampling_prob_ / kLowProb), then replace the coin
+    // flip for kLowProb with a thread local number of tries tries_left_
+    // sampled from the geometric distribution
+    if (sampling_prob_ < kLowProb) {
+      if (tries_left_ == 0) {
+        tries_left_ = sample_geometric();
```

Comment:
you just need 
```
if (tries_left_ == 0){
 tries_left_ = sample_exponential()/sampling_prob_ + .5;
 return true;
} else {...
```
here. Wait time until the next event in poisson process is exponential distribution http://homepage.divms.uiowa.edu/~rdecook/stat2020/notes/ch4_pt3.pdf, and `+.5` is so that you round to nearest when you convert to int instead of round down (you could also use std::round).

Reply:
the issue here is that different callbacks have different sampling_prob_ so I guess we can't use a simple thread local tries_left_ in this case

Reply:
would it be hard to just have per-callback counter? I'm always worried about mixing different distributions.

If we indeed have only a single callback usually, I'm fine landing it as is and following up